pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -24,98 +22,61 @@ var mercuryInfo = capabilities.MustNewCapabilityInfo(
 	"v1.0.0",
 )
 
+const defaultTickerResolutionMs = 1000
+
 // This Trigger Service allows for the registration and deregistration of triggers. You can also send reports to the service.
 type MercuryTriggerService struct {
 	capabilities.CapabilityInfo
-	chans               map[string]chan<- capabilities.CapabilityResponse
-	feedIDsForTriggerID map[string][]mercury.FeedID
-	triggerIDsForFeedID map[mercury.FeedID]map[string]bool
-	mu                  sync.Mutex
-	lggr                logger.Logger
+	tickerResolutionMs int64
+	chans              map[string]chan<- capabilities.CapabilityResponse
+	subscribers        map[string]*subscriber
+	latestReports      map[mercury.FeedID]mercury.FeedReport
+	mu                 sync.Mutex
+	stopCh             services.StopChan
+	wg                 sync.WaitGroup
+	lggr               logger.Logger
+}
+
+var _ services.Service = &MercuryTriggerService{}
+
+type SubscriberConfig struct {
+	FeedIds        []string
+	MaxFrequencyMs int
+}
+
+type subscriber struct {
+	ch         chan<- capabilities.CapabilityResponse
+	workflowID string
+	config     SubscriberConfig
 }
 
 var _ capabilities.TriggerCapability = (*MercuryTriggerService)(nil)
 
-func NewMercuryTriggerService(lggr logger.Logger) *MercuryTriggerService {
+// Mercury Trigger will send events to each subscriber every MaxFrequencyMs (configurable per subscriber).
+// Event generation happens whenever local unix time is a multiple of tickerResolutionMs. Therefore,
+// all subscribers' MaxFrequencyMs values need to be a multiple of tickerResolutionMs.
+func NewMercuryTriggerService(tickerResolutionMs int64, lggr logger.Logger) *MercuryTriggerService {
+	if tickerResolutionMs == 0 {
+		tickerResolutionMs = defaultTickerResolutionMs
+	}
 	return &MercuryTriggerService{
-		CapabilityInfo:      mercuryInfo,
-		chans:               map[string]chan<- capabilities.CapabilityResponse{},
-		feedIDsForTriggerID: make(map[string][]mercury.FeedID),
-		triggerIDsForFeedID: make(map[mercury.FeedID]map[string]bool),
-		lggr:                logger.Named(lggr, "MercuryTriggerService"),
+		CapabilityInfo:     mercuryInfo,
+		tickerResolutionMs: tickerResolutionMs,
+		chans:              map[string]chan<- capabilities.CapabilityResponse{},
+		subscribers:        make(map[string]*subscriber),
+		latestReports:      make(map[mercury.FeedID]mercury.FeedReport),
+		stopCh:             make(services.StopChan),
+		lggr:               logger.Named(lggr, "MercuryTriggerService"),
 	}
 }
 
-type FeedReport struct {
-	FeedID               [mercury.FeedIDBytesLen]byte `json:"feedId"`
-	FullReport           []byte                       `json:"fullReport"`
-	BenchmarkPrice       int64                        `json:"benchmarkPrice"`
-	ObservationTimestamp int64                        `json:"observationTimestamp"`
-}
-
-func (o *MercuryTriggerService) ProcessReport(reports []FeedReport) error {
+func (o *MercuryTriggerService) ProcessReport(reports []mercury.FeedReport) error {
```

Comment:
@bolekk Are we using string-based FeedIDs then? (The reason these are different is because ProcessReport uses a byte based FeedID, but internally we use a string).

Reply:
I think whoever calls ProcessReport can just convert to a string. I didn't notice before that we have two almost identical FeedReport structs, I think it's OK to simplify.